mullvad-ios/src/encrypted_dns_proxy.rs

```diff
@@ -0,0 +1,217 @@
+    NoConfigs,
+}
+
+impl From<Error> for i32 {
+    fn from(err: Error) -> Self {
+        match err {
+            Error::TokioRuntime => -1,
+            Error::BindLocalSocket(_) => -2,
+            Error::GetBindAddr(_) => -3,
+            Error::Forwarder(_) => -4,
+            Error::FetchConfig(_) => -5,
+            Error::NoConfigs => -6,
+        }
+    }
+}
+
+impl EncryptedDnsProxyState {
+    async fn start(&mut self) -> Result<ProxyHandle, Error> {
+        self.fetch_configs().await?;
+        let proxy_configuration = self.next_configuration().ok_or(Error::NoConfigs)?;
+
+        let local_socket = Self::bind_local_addr()
+            .await
+            .map_err(Error::BindLocalSocket)?;
+        let bind_addr = local_socket.local_addr().map_err(Error::GetBindAddr)?;
+        let forwarder = Forwarder::connect(&proxy_configuration)
+            .await
+            .map_err(Error::Forwarder)?;
+        let join_handle = Box::new(tokio::spawn(async move {
+            if let Ok((client_conn, _)) = local_socket.accept().await {
+                let _ = forwarder.forward(client_conn).await;
+            }
+        }));
+
+        Ok(ProxyHandle {
+            context: Box::into_raw(join_handle).cast(),
+            port: bind_addr.port(),
+        })
+    }
+
+    async fn bind_local_addr() -> io::Result<TcpListener> {
+        let bind_addr = SocketAddr::new(Ipv4Addr::UNSPECIFIED.into(), 0);
+        TcpListener::bind(bind_addr).await
+    }
+
+    /// Select a config.
+    /// Always select an obfuscated configuration, if there are any left untried. If no obfuscated
+    /// configurations exist, try plain configurations. The order is randomized due to the hash set
+    /// storing the configurations in a random order.
+    fn next_configuration(&mut self) -> Option<ProxyConfig> {
+        if self.should_reset() {
+            self.reset();
+        }
+
+        // First, try getting an obfuscated configuration, if there exist any.
+        let config = if let Some(obfuscated_config) = self
+            .configurations
+            .difference(&self.tried_configurations)
+            .find(|config| config.obfuscation.is_some())
+            .cloned()
+        {
+            obfuscated_config
+        } else {
+            // If no obfuscated configurations exist, try the next untried configuration.
+            self.configurations
+                .difference(&self.tried_configurations)
+                .next()
+                .cloned()?
+        };
+
+        self.tried_configurations.insert(config.clone());
+        Some(config)
+    }
+
+    /// Fetch a config, but error out only when no existing configuration was there.
+    async fn fetch_configs(&mut self) -> Result<(), Error> {
+        match mullvad_encrypted_dns_proxy::config_resolver::resolve_default_config().await {
+            Ok(new_configs) => {
+                self.configurations = HashSet::from_iter(new_configs.into_iter());
```

Comment:
This part is important for the random order promises we make. If `self.configurations` were not re-created but rather the same `HashSet` was cleared and re-filled, then we would be stuck in trying proxies in the same order every time. Maybe this line should document that it's important for that reason?